### PR TITLE
Adding categorical data support to decision tree

### PR DESCRIPTION
**Pull request by vsanjb8**
## what changes ?
- adding support to categorical data in `split_condition`
- Adding support for conditional split for categorical values using `value in split_condition` check
- removing dependency of order of children
   - in the current version it is expected for first children to be left child and second children to be right child

## unit test changes ?
- change in RAM usage test case as new date members were added to NaiveAdditiveDecisionTree and XGBoostJsonParser classes

## build zip file
[ltr-1.5.9-es7.17.10.zip](https://github.com/practo/elasticsearch-learning-to-rank/files/13446985/ltr-1.5.9-es7.17.10.zip)

### CODE REVIEW

File: src/test/java/com/o19s/es/ltr/ranker/dectree/NaiveAdditiveDecisionTreeTests.java [hunk 1]
```diff
@@ -113,9 +114,11 @@ public void testRamSize() {
                 100, 1000,
                 5, 50, counts);
         long actualSize = ranker.ramBytesUsed();
-        long expectedApprox = counts.splits.get() * (NUM_BYTES_OBJECT_HEADER + Float.BYTES + NUM_BYTES_OBJECT_REF * 2);
+        long expectedApprox = counts.splits.get() * (NUM_BYTES_OBJECT_HEADER + Float.BYTES +
+                RamUsageEstimator.primitiveSizes.get(boolean.class) + NUM_BYTES_ARRAY_HEADER+ NUM_BYTES_OBJECT_REF * 2);
         expectedApprox += counts.leaves.get() * (NUM_BYTES_ARRAY_HEADER + NUM_BYTES_OBJECT_HEADER + Float.BYTES);
         expectedApprox += ranker.size() * Float.BYTES + NUM_BYTES_ARRAY_HEADER;
+        LOG.info("actual size: {}", actualSize);
```
Comment (rahulbu) [inline src/test/java/com/o19s/es/ltr/ranker/dectree/NaiveAdditiveDecisionTreeTests.java:121]: is this required ?

**Review by rahulbu — APPROVED** [review]

**Review by sabhishekgowda — APPROVED** [review]

File: src/main/java/com/o19s/es/ltr/ranker/parser/XGBoostJsonParser.java [hunk 2]
```diff
@@ -153,20 +154,41 @@ private static class SplitParserState {
             PARSER.declareInt(SplitParserState::setNodeId, new ParseField("nodeid"));
             PARSER.declareInt(SplitParserState::setDepth, new ParseField("depth"));
             PARSER.declareString(SplitParserState::setSplit, new ParseField("split"));
-            PARSER.declareFloat(SplitParserState::setThreshold, new ParseField("split_condition"));
             PARSER.declareInt(SplitParserState::setRightNodeId, new ParseField("no"));
             PARSER.declareInt(SplitParserState::setLeftNodeId, new ParseField("yes"));
             PARSER.declareInt(SplitParserState::setMissingNodeId, new ParseField("missing"));
             PARSER.declareFloat(SplitParserState::setLeaf, new ParseField("leaf"));
+            PARSER.declareField((parser, splitParserState, context) -> {
```
Comment (sabhishekgowda) [inline src/main/java/com/o19s/es/ltr/ranker/parser/XGBoostJsonParser.java:161]: can we move this to a function?

Comment (vsanjb8) [reply src/main/java/com/o19s/es/ltr/ranker/parser/XGBoostJsonParser.java:161]: we'll have to create a separate class for it.. as it requires an object extending Parser interface
should we go ahead with that or live with this lambda ?

can create a function which returns this lambda may be

Comment (vsanjb8) [reply src/main/java/com/o19s/es/ltr/ranker/parser/XGBoostJsonParser.java]: have created a separate static class @sabhishekgowda

**Review by sabhishekgowda — APPROVED** [review]

**Review by mayankdharwa — APPROVED** [review]

**Pull request merged**